### ISSUE TITLE
[IDEA] using function calling to integrate more tools

### ISSUE THREAD

**Issue by thinker007** [khoj-ai/khoj#841]
## Describe the feature you'd like
A clear and concise description of what you want to happen. Include any relevant links or screenshots or inspiration.

**Comment by sabaimran** [comment]
hey @thinker007 , could you describe more of what you're looking for? Do you want another tool similar to Dify? https://dify.ai/blog/dify-ai-workflow

**Comment by thinker007** [comment]
> hey @thinker007 , could you describe more of what you're looking for? Do you want another tool similar to Dify? https://dify.ai/blog/dify-ai-workflow

yes dify

**Comment by DavidLMS** [comment]
I think it would be beneficial to allow agents to define custom actions in a more generic manner, similar to OpenAI's approach with function calling in GPTs. This would offer greater flexibility for integrating various workflows and tools, extending beyond the built-in options currently available.

**Comment by thinker007** [comment]
> I think it would be beneficial to allow agents to define custom actions in a more generic manner, similar to OpenAI's approach with function calling in GPTs. This would offer greater flexibility for integrating various workflows and tools, extending beyond the built-in options currently available.

or let devs integrate tools made from dify , we don't need second dify, what we need is integrate dify tools workflow into khoj agents

**Comment by thinker007** [comment]
By the way  dify can turn tools into rest https api, if khoj can let devs integrate general http service it will be great!

**Comment by felipemeres** [comment]
I agree, even a simple implementation of a tool framework that lets users setup different request formats to api endpoints and call them with assigned /commands would be great.

**Comment by sabaimran** [comment]
These sound like really great ideas! Can you give me an idea of some tools you'd integrate? What would they do, how would you use them, how would you handle the responses? Be as detailed as possible, as that would help us design the feature. 

I think the ideal tool might be something that can be called via an http request, like @thinker007 mentioned, where of course you can define the input parameters / body structure, and how to handle the response. 

Unless the tool can just be text in - text out from an LLM.

I think custom code execution would probably be out of scope.

**Comment by thinker007** [comment]
> These sound like really great ideas! Can you give me an idea of some tools you'd integrate? What would they do, how would you use them, how would you handle the responses? Be as detailed as possible, as that would help us design the feature.
> 
> I think the ideal tool might be something that can be called via an http request, like @thinker007 mentioned, where of course you can define the input parameters / body structure, and how to handle the response.
> 
> Unless the tool can just be text in - text out from an LLM.
> 
> I think custom code execution would probably be out of scope.

Yes out there is plenty of agent tools just like dify, in khoj we just need a good way to integrate them

**Comment by felipemeres** [comment]
The [External Knowledge](https://docs.dify.ai/guides/knowledge-base/external-knowledge-api-documentation) feature of Dify could be a good reference. Essentially you setup the api endpoint and any agent can call this function to retrieve its output as context.

**Comment by ferhimedamine** [comment]
@felipemeres's point about the External Knowledge API pattern is the right direction. One thing I've learned building a memory server for agents: the biggest design win is keeping memory as a plain HTTP tool the agent calls, rather than wiring retrieval logic into the agent framework itself. The agent stores observations via POST and recalls context via GET with a query — same as any other tool call. This means you get session-scoped recall, importance ranking, and hybrid search (BM25 + vector) without the orchestration layer needing to know anything about embeddings or retrieval strategy.

The practical benefit for something like Khoj's tool integration: if tool responses are worth remembering across conversations, the agent just calls a `/store` endpoint after receiving the tool output, and later recalls it via `/recall` with the user's query. No changes to the tool framework needed — it's just another HTTP tool in the list.

I've been working on exactly this pattern — the deploy setup is a single `docker-compose.yml` if anyone wants to experiment with it: https://github.com/dakera-ai/dakera-deploy/blob/main/docker/docker-compose.yml. It runs a self-hosted memory server the agent calls as an external tool, which is basically what this thread is converging on for the knowledge retrieval case.
